
**I tried to find the model that worked best using different models in the Boston housing dataset.**

In [2]:
import tensorflow as tf

In [3]:
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.boston_housing.load_data()

In [10]:
x_train[:5]

array([[1.23247e+00, 0.00000e+00, 8.14000e+00, 0.00000e+00, 5.38000e-01,
        6.14200e+00, 9.17000e+01, 3.97690e+00, 4.00000e+00, 3.07000e+02,
        2.10000e+01, 3.96900e+02, 1.87200e+01],
       [2.17700e-02, 8.25000e+01, 2.03000e+00, 0.00000e+00, 4.15000e-01,
        7.61000e+00, 1.57000e+01, 6.27000e+00, 2.00000e+00, 3.48000e+02,
        1.47000e+01, 3.95380e+02, 3.11000e+00],
       [4.89822e+00, 0.00000e+00, 1.81000e+01, 0.00000e+00, 6.31000e-01,
        4.97000e+00, 1.00000e+02, 1.33250e+00, 2.40000e+01, 6.66000e+02,
        2.02000e+01, 3.75520e+02, 3.26000e+00],
       [3.96100e-02, 0.00000e+00, 5.19000e+00, 0.00000e+00, 5.15000e-01,
        6.03700e+00, 3.45000e+01, 5.98530e+00, 5.00000e+00, 2.24000e+02,
        2.02000e+01, 3.96900e+02, 8.01000e+00],
       [3.69311e+00, 0.00000e+00, 1.81000e+01, 0.00000e+00, 7.13000e-01,
        6.37600e+00, 8.84000e+01, 2.56710e+00, 2.40000e+01, 6.66000e+02,
        2.02000e+01, 3.91430e+02, 1.46500e+01]])

In [12]:
y_train[:5]

array([15.2, 42.3, 50. , 21.1, 17.7])

In [13]:
print(len(x_train))
print(len(x_test))

404
102


In [4]:
from sklearn.model_selection import train_test_split

x_test,x_val,y_test,y_val = train_test_split(x_test,y_test,test_size=0.5,random_state = 58)

In [5]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()

x_train = sc.fit_transform(x_train) 
x_test = sc.transform(x_test)
x_val = sc.transform(x_val)

In [21]:
x_train.shape

(404, 13)

In [53]:
ann_model = tf.keras.models.Sequential()

ann_model.add(tf.keras.layers.Dense(units = 128,activation="relu"))
ann_model.add(tf.keras.layers.Dropout((0.3)))
ann_model.add(tf.keras.layers.Dense(units = 128,activation="relu"))
ann_model.add(tf.keras.layers.Dropout((0.3)))

ann_model.add(tf.keras.layers.Dense(units = 1 , activation="linear"))

ann_model.compile(optimizer="adam" , loss = "mean_squared_error",metrics = ["mae"])

fitted_ann_model = ann_model.fit(x_train,y_train,epochs=50,batch_size= 4 ,validation_data=(x_val,y_val))

Epoch 1/50
101/101 [==============================] - 2s 5ms/step - loss: 276.4594 - mae: 13.7703 - val_loss: 51.2996 - val_mae: 5.4740
Epoch 2/50
101/101 [==============================] - 0s 3ms/step - loss: 44.0062 - mae: 4.8520 - val_loss: 27.3342 - val_mae: 4.0448
Epoch 3/50
101/101 [==============================] - 0s 4ms/step - loss: 30.9466 - mae: 4.0567 - val_loss: 26.9236 - val_mae: 3.8899
Epoch 4/50
101/101 [==============================] - 0s 4ms/step - loss: 29.2249 - mae: 3.7814 - val_loss: 27.6585 - val_mae: 3.7472
Epoch 5/50
101/101 [==============================] - 0s 4ms/step - loss: 21.8297 - mae: 3.4325 - val_loss: 29.4606 - val_mae: 3.8176
Epoch 6/50
101/101 [==============================] - 0s 4ms/step - loss: 23.0356 - mae: 3.5169 - val_loss: 30.7513 - val_mae: 3.6410
Epoch 7/50
101/101 [==============================] - 0s 4ms/step - loss: 19.8428 - mae: 3.3236 - val_loss: 33.7407 - val_mae: 3.7534
Epoch 8/50
101/101 [==============================] - 0s 3ms

In [54]:
ann_model.evaluate(x_test,y_test)

2/2 [==============================] - 0s 9ms/step - loss: 12.2702 - mae: 2.6966


[12.27016544342041, 2.6966166496276855]

In [55]:
from sklearn import svm

In [60]:
svm_model = svm.SVR()
svm_model.fit(x_train,y_train)
predict_svm = svm_model.predict(x_test)

In [61]:
from sklearn.metrics import mean_absolute_error
mean_absolute_error(y_test,predict_svm)

3.6536811561464533

In [11]:
y_train = y_train.reshape(-1,1)
y_test = y_test.reshape(-1,1)

In [22]:
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
import numpy as np
from sklearn.metrics import mean_absolute_error

poly_degree = 2

poly_features = PolynomialFeatures(degree=poly_degree)

x_train_poly = poly_features.fit_transform(x_train)
x_test_poly = poly_features.transform(x_test)

poly_reg = LinearRegression()
poly_reg.fit(x_train_poly, y_train)

LinearRegression()

In [23]:
predict_poly = poly_reg.predict(x_test_poly)
mean_absolute_error(y_test,predict_poly)

2.7496323529411764

1.   Ann model --> mean absolute error = 2.70
2.   Polynomial LR model --> mean absolute error = 2.75
3.   Svm model --> mean absolute error = 3.65